tools/ncf_constraints.py

```diff
@@ -0,0 +1,90 @@
+import re
+import subprocess
+import json
+import os.path
+import shutil
+import sys
+import os
+import codecs
+
+
+### Constraint checking function 
+
+def from_list( parameter, accepted_result):
+  return parameter in accepted_result
+
+def max_length( parameter, max_size):
+  return len(parameter) <= max_size
+
+def min_length( parameter, min_size):
+  return len(parameter) >= min_size
+
+def match_regex(parameter, regex):
+  match = re.search(regex, parameter)
+  return match is not None
+
+def not_match_regex(parameter, regex):
+  return not match_regex(parameter, regex)
+
+def check_parameter(parameter, constraint):
+  """Checks that a parameter is ok with the constraint passed as parameter"""
+
+  result = True
+  errors = [];
+
+  # First check 'string' constraints
+  if not constraint.get("allow_empty_string", False):
+    if match_regex(parameter, "^$"):
+      errors.append("allow_empty_string")
+      result = False
+
+  if not constraint.get("allow_whitespace_string", False):
+    if match_regex(parameter, r'^\s') or match_regex(parameter, r'\s+$'):
+      errors.append("allow_whitespace_string")
+      result = False
+
+  # Check if in list of accepted values
+  if "list" in constraint:
+    if not from_list(parameter, constraint["list"]):
+      errors.append("list")
+      result = False
+
+  # Check regexs constraints
+  if "regex" in constraint:
+    if not match_regex(parameter, constraint["regex"]):
+      errors.append("regex")
+      result = False
+
+  if "not_regex" in constraint:
+    if not not_match_regex(parameter, constraint["not_regex"]):
+      errors.append("not_regex")
+      result = False
+
+  # Check length constraints
+  if "min_length" in constraint:
+    if not min_length(parameter, constraint["min_length"]):
+      errors.append("min_length")
+      result = False
+
+  if "max_length" in constraint:
+    if not max_length(parameter, constraint["max_length"]):
+      errors.append("max_length")
+      result = False
+  
```

Comment:
This touches on the specs for this feature, which were done a while ago now. We thought back then that it would be sufficient for the first implementation to have a list of *required* constraints only. This makes sense for many combinaisons (min_length + max_length + matches regex + forbidden regex, for example).

The list can be useful in conjuction with other constraints to *relax* the strictness: for example ```constraints: 'list': {'a', 'b', 'c'}, 'allow_whitespace_string': true``` would allow values a, b or c, but also tolerate leading/trailing whitespace.